proposals/nnnn-variadic-types.md

```diff
@@ -0,0 +1,239 @@
+# Allow Generic Types to Abstract Over Packs
+
+* Proposal: [SE-NNNN](nnnn-variadic-types.md)
+* Authors: [Slava Pestov](https://github.com/slavapestov), [Holly Borla](https://github.com/hborla)
+* Review Manager: TBD
+* Upcoming Feature Flag: `VariadicGenerics`
+* Previous Proposal: [SE-0393](0393-parameter-packs.md)
+
+## Introduction
+
+Previously [SE-0393](0393-parameter-packs.md) introduced type parameter packs and several related concepts, allowing generic function declarations to abstract over a variable number of types. This proposal generalizes these ideas to generic type declarations.
+
+## Motivation
+
+Generic type declarations that abstract over a variable number of types arise naturally when attempting to generalize common algorithms on collections. For example, a lazy `ZipSequence` might be generic over two sequences. It would be possible to declare a `ZipSequence` type which presents the elements of a fixed list of sequences as a sequence of tuples:
+
+```swift
+struct ZipSequence<each S: Sequence>: Sequence {
+  typealias Element = (repeat each S.Element)
+
+  let seq: (repeat each S)
+
+  func makeIterator() -> Iterator {
+    return Iterator(iter: (repeat (each seq).makeIterator()))
+  }
+
+  struct Iterator: IteratorProtocol {
+    typealias Element = (repeat each S.Element)
+
+    var iter: (repeat each S.Iterator)
+
+    mutating func next() -> Element? {
+      return ...
+    }
+  }
+}
+```
+
+## Proposed solution
+
+In the generic parameter list of a generic type, the `each` keyword declares a generic parameter pack, just like it does in the generic parameter list of a generic function. The types of stored properties can contain pack expansion types, as in `let seq` and `var iter` above.
+
+## Detailed design
+
+Swift has the following kinds of generic type declarations:
+
+- Structs
+- Enums
+- Classes (and actors)
+- Type aliases
+
+A generic type is _variadic_ if it directly declares a type parameter pack with `each`, or if it is nested inside of another variadic type. In this proposal, structs, classes, actors and type aliases can be variadic. Enums will be addressed in a follow-up proposal.
```

Comment:
Might be nice to have a brief mention of why they're being treated separately, like we have for classes below?